pkg/block/fetcher.go

```diff
@@ -566,4 +573,51 @@ func (f *ConsistencyDelayMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta,
 			delete(metas, id)
 		}
 	}
+
+	return nil
+}
+
+// IgnoreDeletionMarkFilter is a MetaFetcher that filters blocks that are marked for deletion; delay duration before current time.
+// The delay duration is to make sure that the replacement block can be fetched before we filter out the old block.
+type IgnoreDeletionMarkFilter struct {
+	logger          log.Logger
+	delay           time.Duration
+	bkt             objstore.BucketReader
+	deletionMarkMap map[ulid.ULID]*metadata.DeletionMark
+}
+
+// NewIgnoreDeletionMarkFilter creates IgnoreDeletionMarkFilter.
+func NewIgnoreDeletionMarkFilter(logger log.Logger, bkt objstore.BucketReader, delay time.Duration) *IgnoreDeletionMarkFilter {
+	return &IgnoreDeletionMarkFilter{
+		logger: logger,
+		bkt:    bkt,
+		delay:  delay,
+	}
+}
+
+// DeletionMarkBlocks returns block ids that were marked for deletion.
+func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks() map[ulid.ULID]*metadata.DeletionMark {
+	return f.deletionMarkMap
+}
+
+// Filter filters out blocks that are marked for deletion; delay duration before current time.
+// It also returns the blocks that can be deleted since they were uploaded delay duration before current time.
+func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) error {
+	f.deletionMarkMap = make(map[ulid.ULID]*metadata.DeletionMark)
+
+	for id := range metas {
+		deletionMark, err := metadata.ReadDeletionMark(ctx, f.bkt, f.logger, id.String())
+		if err == metadata.ErrorDeletionMarkNotFound {
+			continue
+		}
+		if err != nil {
+			return err
```

Comment:
What if deletion mark is unparsable because it's being in upload? :thinking:

Reply:
Lets discuss how we can detect unparsable deletion-mark.json over our call.

Reply:
What would that function do? Maybe let's return some sentinel error, check for it, and then log + continue the loop? Because not just `Unmarshal` can fail, it can also have a wrong version number.

Reply:
I have created my own error type `ErrorUnmarshalDeletionMark` and I am returning that in case that the unmarshal fails. I think that is the right way to deal with this. Please open a new thread if you have other thoughts. Thanks.